### ISSUE TITLE
Pyspark intrinsic requirements in segments that shouldn't exist

### ISSUE THREAD

**Issue by mauro-lanza** [sqlfluff/sqlfluff#5439]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

LateralViewClauseSegment must be written just below a FROM clause, when moving it to after some joins it gets parsing error, this is an issue for lateral views that explode a column coming from the second table since it need to be written after the join.

ClusterByClauseSegment must have a FROM clause in the query for it to be valid so something like below would also cause errors:
```sql
select CURRENT_DATE() as p_data_date
cluster by p_data_date
```
this is also wrong since the above expression would work correctly.

### Expected Behaviour

LateralViewClauseSegment should be allowed after any from or join expression.
ClusterByClauseSegment should be allowed on queries without from statement.

### Observed Behaviour

what has been written on what happened

### How to reproduce

LateralViewClauseSegment this will give two errors, one on the lateral view and the other one when using the alias 'things' from the lateral view column in the select

```sql
SELECT
    t1.column1,
    CAST(GET_JSON_OBJECT(things, '$.percentage') AS DECIMAL(16, 8)
    ) AS ptc
FROM table1 AS t1
LEFT JOIN table2 AS t2
    ON c.column1 = p.column1
    AND t2.type = 'SOMETHING'
  LATERAL VIEW OUTER EXPLODE(t2.column2) AS things;
```

ClusterByClauseSegment:

```sql
SELECT CURRENT_DATE() AS p_data_date
CLUSTER BY p_data_date;
```

### Dialect

dialect = sparksql
templater = jinja

### Version

2.3.5

### Configuration

[sqlfluff]
dialect = sparksql
templater = jinja
max_line_length = 0
exclude_rules = ST06
fix_even_unparsable = true

[sqlfluff:indentation]
allow_implicit_indents = True
indented_on_contents = False
indented_ctes = True

[sqlfluff:layout:type:alias_expression]
spacing_before = align
align_within = select_clause
align_scope = bracketed

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper
[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = upper
[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = upper
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = upper
[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = upper

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by mauro-lanza** [comment]
It has been commented here before https://github.com/sqlfluff/sqlfluff/issues/5325#issuecomment-1776939385
